### ISSUE TITLE
integrate(atan(tan(x)), x) etc. gives wrong anti-derivative

### ISSUE THREAD

**Issue by andh0g** [sympy/sympy#20898]
There is some code in the method doit of the class Integral (line 605--626 in sympy/integrals/integrals.py, latest version on master at time of writing) to handle discontinuous antiderivatives.  I have found three examples that gives the wrong answer with the current code.

Example 1:

    from sympy import integrate, atan, tan, cos, Symbol
    x = Symbol('x')
    integrate(atan(tan(x)), x)

gives

    (atan(tan(x)) + pi*floor((x - pi/2)/pi))**2/2

which is correct if the floor function is removed since there is no discontinuity to remove.  With the floor function, the answer is equivalent to `(x-pi)**2/2`.

Example 2:

    integrate(atan(tan(x))**2, x)

gives

    (atan(tan(x)) + pi*floor((x - pi/2)/pi))**3/3

which would be correct if the floor function was added outside of the
power (of course with a different coefficient).  As it is now, the
answer is equivalent to `(x-pi)**3/3`.

Example 3:

    integrate((cos(2*x))/(2+cos(x)), x).simplify()

gives

    -4*x + 2*sin(x) + 14*sqrt(3)*atan(sqrt(3)*tan(x/2)/3)/3 + 28*sqrt(3)*pi*floor(x/(2*pi) - 1/2)/3

This example gives a rather messy answer but after simplification we
can see that the coefficient of the floor function is twice as big as
it should be.  I suspect this is due to some recursion causing the
floor function to be added twice.

**Comment by hyadav2k** [comment]
@andh0g, for `integrate(atan(tan(x)), x)`, wolfram is giving `-x^2/2 + x tan^(-1)(tan(x)) + constant`. Even solving manually using integration by parts, it comes same as wolfram's answers, thus, Sympy's answer is incorrect.

The answer is evaluated at:
https://github.com/sympy/sympy/blob/9e8f62e059d83178c1d8a1e19acac5473bdbf1c1/sympy/integrals/integrals.py#L612-L613

**Comment by Maelstrom6** [comment]
Wolfram's answer is also wrong. Integrals of functions that do not contain dirac deltas are all continuous. Wolfram's answer is not continuous and so it will produce incorrect definite integrals.

> The answer is evaluated at:
> https://github.com/sympy/sympy/blob/9e8f62e059d83178c1d8a1e19acac5473bdbf1c1/sympy/integrals/integrals.py#L612-L613

Thanks for finding that line. Saves everyone a lot of work. We will have to decide what to replace it with.

**Comment by jksuom** [comment]
`tan(x)` is a periodic function of period `pi`. Therefore `atan(tan(x)l` is also periodic. Its value is `x` in the interval `(-pi/2, pi/2)` and it is discontinuous because its.limits at the endpoints are different. The purpose of the `floor` term is to remove this discontinuity.

**Comment by Maelstrom6** [comment]
![Screenshot_20210208_120735.jpg](https://user-images.githubusercontent.com/45897212/107205211-3ff06800-6a06-11eb-8521-dab75aa21e20.jpg)

Wolfram's answer is discontinuous which is incorrect. If it took this as the correct solution then the integral from 0 to pi will be -pi according to Wolfram but visually, it is 0.

**Comment by almighty123789** [comment]
Actually, you **cannot** find an antiderative for the `atan(tan(x)) ` function. This is because the it involves floor in it's simlification i.e 
` atan(tan(x)) = x-pi * ( floor( x/pi-1/2)+1)`, since the floor function depends on the limit, you cannot find an exact antideravative for it, since it depends on the limit itself, withtout the limits one cannot simplify floor function
currently sympy's integral module has no method for computing integrals of floor functions, thus to solve this problem, we would first require a way to integrate floor function.

**Comment by Maelstrom6** [comment]
atan(tan(x)) does have an anitderivaive. It is atan(tan(x))^2/2 as @andh0g has stated in their post:

![Screenshot_20210208_124349.jpg](https://user-images.githubusercontent.com/45897212/107209377-6533a500-6a0b-11eb-958d-31c034561134.jpg)

As we can see, the anitderivaive is continuous and trial values like 0, pi/2 and pi make sense. I'm sure you could always differentiate the anitderivaive by hand to check that it is a solution.

**Comment by almighty123789** [comment]
Yeah,you are right, this  is the correct antideravative, moreover this antiderative can actually be used to find the value of definite integrals involving atan

**Comment by oscarbenjamin** [comment]
I wonder if we should just have antiderivatives that are only locally correct. Perhaps the logic for using an antiderivative of a discontinuous integrand should be handled in the definite integration code rather than by trying to find an antiderivative that can account for the discontinuities by introducing continuous but non-differentiable corrections.

The antiderivative for `atan(tan(x))` works for definite integrals over real intervals but it is not a true "antiderivative" since it isn't differentiable everywhere. Maybe we should just accept that the same is true for other cases and not bother trying to smooth over with floor.

**Comment by andh0g** [comment]
This is not only a problem for discontinuous integrands as my example 2 shows.  The integrand `atan(tan(x))**2` is continuous but needs a correction in the antiderivative to be continuous.

I think this is a very hard problem to solve so the suggestion of @oscarbenjamin not to try to fix the problem with floor functions seems like a reasonable choice.  It is easy for me as a user to scan the result for things needing attention.  However, is it easier to solve for definite integrals?  It is usually  hard for me to check that the number I get back is correct.

I also notice that the `atan(tan(x)).diff(x)` returns a constant 1 while it refuses to differentiate a floor function, so this is not just a problem for integration.

**Comment by oscarbenjamin** [comment]
I think in general it is not always possible to create antiderivative expressions that are everywhere valid for definite integrals meaning that there will always be some cases that have to be handled in the definite integration code. Discussion and examples here:
http://www.stat.rice.edu/~dobelman/notes_papers/math/Integration.definites.davenport.pdf

More generally I would just like it if the intended specification for what integrate should return was clearly documented. Looking in the docs I don't see a clear definition:
https://docs.sympy.org/latest/modules/integrals/integrals.html#sympy.integrals.integrals.integrate

**Comment by oscarbenjamin** [comment]
A simple troublesome example:
```python
In [29]: integrate(1/x**2, x)
Out[29]: 
-1 
───
 x 

In [30]: integrate(1/x**2, (x, -1, 1))
Out[30]: ∞
```

**Comment by sylee957** [comment]
I have also stumbled across this problem, but I think that this is a natural consequence of Risch algorithm that cannot be easily resolved.
There is an article about this in https://blog.wolfram.com/2008/01/19/mathematica-and-the-fundamental-theorem-of-calculus/

And it also says that it's not possible to make a continuous antiderivative for whole complex plane for some functions
> Moreover, if a meromorphic integrand h(z) has simple poles in the complex plane, it is impossible to choose an antiderivative H(z) continuous along every imaginable path in the complex plane—because of branch cuts in H(z).

Although I'm not sure if this discourages any attempt to make continuous antiderivative for integration along real axis, but it's worthwhile that it just needs a separate scope of implementation.

I think that `integrate(f, x)` and `Integrate(f, (x, a, x))` are different problem (analytic definition of antiderivative and algebraic definition of antiderivative) and the latter one is significantly difficult to compute than the former one.

**Comment by che2712** [comment]
While typing in examples, this is what i got and I think the problem is with the functioning of Integrate(f, (x, a, b)) as follows:
>>> integrate(1/x**2, (x, -1))
1

>>> integrate(1/x**2, (x, 1))
−
1


>>> integrate(1/x**2, (x, -1, x))
−1−1/x

**Comment by almighty123789** [comment]
> A simple troublesome example:
> 
> ```python
> In [29]: integrate(1/x**2, x)
> Out[29]: 
> -1 
> ───
>  x 
> 
> In [30]: integrate(1/x**2, (x, -1, 1))
> Out[30]: ∞
> ```

@oscarbenjamin  what exactly is the trouble with this integral?

**Comment by oscarbenjamin** [comment]
> what exactly is the trouble with this integral?

If you substitute the limits into the expression for the antiderivative then you will not get the same result as the definite integral.

**Comment by almighty123789** [comment]
> > what exactly is the trouble with this integral?
> 
> If you substitute the limits into the expression for the antiderivative then you will not get the same result as the definite integral.

Actually the Integral  ``integrate(1/x**2, (x, -1, 1))`` , an **Improper Integral** and thus expecting that Fundamental Theorem of  calculus holds on this is wrong .

**Comment by jksuom** [comment]
The fundamental theorem cannot be applied as there is no antiderivative in the interval `(-1, 1)` (because of the singularity at the origin).

**Comment by almighty123789** [comment]
> The fundamental theorem cannot be applied as there is no antiderivative in the interval `(-1, 1)` (because of the singularity at the origin).

Yeah exactly what I wanted to convey @jksuom , Thank You.

**Comment by sylee957** [comment]
I think that the floor term should only be added only when weierstrass substitution is used for `sin(x), cos(x)`
https://doi.org/10.1145/174603.174409

The implementation looks different than the paper above,
https://github.com/sympy/sympy/pull/13808
and I now get doubts whether substituting `atan` atomwise was a solution compatible with that.

But the big problem is whether we need to disintegrate weierstrass substitutions from `heurisch`,
but I think that the examples in https://github.com/sympy/sympy/pull/13808 indicates that it shouldn't go deeper than a simple rational or square root compositions of trigonometric functions, so the floor correction could only be applied for them
